File: part_2/swift_solution/Sources/ExchangeLib/OrderBook.swift

class OrderBook
{
    var buyOrders: PriorityQueue<Order> = PriorityQueue<Order>(sort: { (lhs, rhs) -> Bool in
       
        assert(lhs.quantity >= 0)
        assert(lhs.quantity >= 0)
        
        if rhs.price == lhs.price {
            return lhs.generation < rhs.generation;
        }
        
        return rhs.price < lhs.price
    })
    
    var sellOrders: PriorityQueue<Order> = PriorityQueue<Order>(sort: { (lhs, rhs) -> Bool in
        
        assert(lhs.quantity < 0)
        assert(rhs.quantity < 0)
        
        if lhs.price == rhs.price {
            return lhs.generation < rhs.generation
        }
        
        return lhs.price < rhs.price
    })
    
    public func execute(order: Order) -> [Trade]
    {
        var trades : [Trade] = []

        if order.quantity > 0
        {
            buyOrders.enqueue(order)
        }

        if order.quantity < 0
        {
            sellOrders.enqueue(order)
        }
        
        while !buyOrders.isEmpty && !sellOrders.isEmpty
        {
            let buy = buyOrders.peek()!
            let sell = sellOrders.peek()!
        
            if buy.price < sell.price {
                break
            }
            
            let quantity = min(buy.remainingQuantity, sell.remainingQuantity)
            let price = buy.generation < sell.generation ? buy.price : sell.price

            let trade = Trade(buyer: buy.participant,
                              seller: sell.participant,
                              instrument: buy.instrument,
                              quantity: quantity,
                              price: price)
                        
            trades.append(trade)

            buy.fill(amount:quantity)
            sell.fill(amount:quantity)

            if buy.remainingQuantity == 0 {
                let _ = buyOrders.dequeue()
            }

            if sell.remainingQuantity == 0 {
                let _ = sellOrders.dequeue()
            }
        }
        
        return trades
    }
}

File: part_2/csharp_solution/Exchange/OrderBook.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Exchange;

class BuyComparer : IComparer<Order.PriorityKey>
{
    public int Compare(Order.PriorityKey left, Order.PriorityKey right)
    {
        var result = right.Price.CompareTo(left.Price);

        if (result == 0) {
            return left.Generation.CompareTo(right.Generation);
        }

        return result;
    }
}

class SellComparer : IComparer<Order.PriorityKey>
{
    public int Compare(Order.PriorityKey left, Order.PriorityKey right)
    {
        var result = left.Price.CompareTo(right.Price);

        if (result == 0) {
            return left.Generation.CompareTo(right.Generation);
        }

        return result;
    }
}

public class OrderBook
{
    public Trade[] Execute(Order order)
    {
        if (order.Quantity > 0) {
            BuyOrders.Enqueue(order, order.Priority);
        }    
        else {
            SellOrders.Enqueue(order, order.Priority);
        }

        var trades = new List<Trade>();

        while (BuyOrders.Count > 0 && SellOrders.Count > 0) {

            var buy = BuyOrders.Peek();
            var sell = SellOrders.Peek();

            if (buy.Price < sell.Price) {
                break;
            }

            var quantity = Math.Min((buy.RemainingQuantity), sell.RemainingQuantity);
            var price = buy.Generation < sell.Generation ? buy.Price : sell.Price;

            var trade = new Trade(buy.Participant, sell.Participant, buy.Instrument, quantity, price);

            trades.Add(trade);

            buy.Fill(quantity);
            sell.Fill(quantity);

            if (buy.RemainingQuantity == 0) {
                BuyOrders.Dequeue();
            }

            if (sell.RemainingQuantity == 0) {
                SellOrders.Dequeue();
            }
        }
        
        return trades.ToArray();
    }

    public PriorityQueue<Order, Order.PriorityKey> BuyOrders = new (new BuyComparer());
    public PriorityQueue<Order, Order.PriorityKey> SellOrders = new (new SellComparer());

}
